### PR TITLE
Pass the ptls as an argument to specsig calls

### LINKED ISSUE #50068: Performance regression for recursive fib
Hi,
I raised this on discourse but did not get a feedback. Thought I would raise it here. Maybe I am doing something wrong? (Fyi I am on windows)

Code:
```julia
fib(n::Int)::Int = begin
    if n == 0 || n == 1
        return n
    elseif n == 2 
        return 1
    else
        return fib(n - 1) + fib(n - 2)
    end
end
```

In Julia 1.9
`@btime fib(40) # 380.588ms`

In Master commit f407a4cac3
`@btime fib(40) # 1.034s`

Thank you

### PR DESCRIPTION
**Pull request by gbaraldi**
The safepoint at function entry made it so that every function call did a relatively expensive load from the PTLS, we can instead pass the PTLS as an argument to functions making it significantly cheaper.
Also use the swift calling conventions, that together with the `swiftself` attribute makes it so it's very likely the argument is kept in a register between calls.

TODO : What attributtes should we add to that argument? (NonNull,NoAlias etc?)
Fixes https://github.com/JuliaLang/julia/issues/50068

### CODE REVIEW

**Comment by vchuravy** [comment]
Neat. So we are using the swiftself argument to pass the pgcstack? I almost wonder if we should instead use that to pass current_task.

**Comment by gbaraldi** [comment]
Yeah, 
it looks like
```llvm
define swiftcc i64 @julia_fib_133({}*** swiftself %0, i64 signext %1) #0 !dbg !4 {
top:
  %ptls_field3 = getelementptr inbounds {}**, {}*** %0, i64 2
  %2 = bitcast {}*** %ptls_field3 to i64***
  %ptls_load45 = load i64**, i64*** %2, align 8, !tbaa !8
  %3 = getelementptr inbounds i64*, i64** %ptls_load45, i64 2
  %safepoint = load i64*, i64** %3, align 8, !tbaa !12
  fence syncscope("singlethread") seq_cst
  %4 = load volatile i64, i64* %safepoint, align 8, !dbg !14
  fence syncscope("singlethread") seq_cst
  %5 = icmp ult i64 %1, 3, !dbg !15
  br i1 %5, label %switch.lookup, label %L12, !dbg !15

switch.lookup:                                    ; preds = %top
  %switch.gep = getelementptr inbounds [3 x i64], [3 x i64]* @switch.table.julia_fib_133, i64 0, i64 %1, !dbg !15
  %switch.load = load i64, i64* %switch.gep, align 8, !dbg !15
  br label %common.ret, !dbg !15

common.ret:                                       ; preds = %switch.lookup, %L12
  %common.ret.op = phi i64 [ %10, %L12 ], [ %switch.load, %switch.lookup ]
  ret i64 %common.ret.op, !dbg !16

L12:                                              ; preds = %top
  %6 = add i64 %1, -1, !dbg !17
  %7 = call swiftcc i64 @julia_fib_133({}*** nonnull swiftself %0, i64 signext %6), !dbg !20
  %8 = add i64 %1, -2, !dbg !17
  %9 = call swiftcc i64 @julia_fib_133({}*** nonnull swiftself %0, i64 signext %8), !dbg !20
  %10 = add i64 %9, %7, !dbg !21
  br label %common.ret
}
```
we track the `pgcstack` on codegen so I used that.

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50093#issuecomment-1581090075) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6a68b49_vs_13e40af/report.md).

File: src/llvm-ptls.cpp [hunk 1]
```diff
@@ -314,6 +314,19 @@ bool LowerPTLS::run(bool *CFGModified)
...
+                if (arg->hasSwiftSelfAttr()){
+                    pgcstack = &*arg;
+                    break;
+                }
+            }
+            if (pgcstack) {
+                call->replaceAllUsesWith(pgcstack);
+                call->eraseFromParent();
+                continue;
+            }
```
Comment (pchintalapudi) [inline src/llvm-ptls.cpp:329]: If we're going to have the pgcstack as an argument is there any point in not lowering the function in codegen itself?

**Comment by maleadt** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/50093#issuecomment-1594185500) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c0def99_vs_0fb86dd/report.html).

**Comment by gbaraldi** [comment]
GPUCompiler failed but that's kind of expected, it needs to set the correct codegen param

**Pull request merged**